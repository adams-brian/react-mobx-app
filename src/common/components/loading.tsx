import { inject, observer } from 'mobx-react';
import * as React from 'react';

interface IState {
  error: boolean;
  loaded: boolean;
  message: string;
  timeoutFired: boolean;
}

export default <TState extends {}, TExternalProps extends {}>(storeName: string, load: (state: TState) => Promise<void>) =>
  <TProps extends {}>(Component: React.ComponentType<TProps>) => {

    const ObserverComponent = observer(Component);
    
    return inject(storeName)(
      class extends React.Component<TExternalProps, IState> {
        public static displayName = `loading(${Component.displayName || Component.name})`;
        private timeout: NodeJS.Timer;

        constructor(props: TExternalProps) {
            super(props);
            this.state = {
                error: false,
                loaded: false,
                message: "",
                timeoutFired: false
            };
        }

        public componentDidMount() {
          this.timeout = setTimeout(() => this.setState({timeoutFired: true}), 500);
          load(this.props[storeName])
            .then(() => this.setState({loaded: true}))
            .catch((err) => this.setState({error: true, message: err.message}));
        }

        public componentWillUnmount() {
          clearTimeout(this.timeout);
        }

        public render() {
          if (this.state.loaded) {
            return (
              <ObserverComponent {...this.props} />
            );
          }
          else if (this.state.error) {
            return (
              <div>Error: {this.state.message}</div>
            )
          }
          else if (this.state.timeoutFired) {
            return (
              <div>Loading...</div>
            )
          }
          else {
            return null;
          }
        }
      }
    );
  }
