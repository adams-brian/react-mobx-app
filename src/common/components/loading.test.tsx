import { mount } from 'enzyme';
import { observable } from 'mobx';
import { Provider } from 'mobx-react';
import * as React from 'react';
import * as sinon from 'sinon';

import loading from './loading';

class SimpleState {
  @observable public message: string = '';
  public delay: number = 0;
  public shouldReject: boolean = false;
  public load(): Promise<void> {
    return new Promise((resolve, reject) => {
      setTimeout(
        () => {
          if (!this.shouldReject) {
            this.message = 'loaded';
            resolve();
          }
          else {
            reject(new Error('error message'));
          }
        },
        this.delay
      );
    });
  }
}

const SimpleComponent = (props: {theState: SimpleState}) =>
  <div>message: {props.theState.message}</div>

const Wrapper = loading<SimpleState, {}>('theState', theState => theState.load())(
  SimpleComponent
);

describe("loading", () => {

  let state: SimpleState;
  let element: JSX.Element;
  let clock: sinon.SinonFakeTimers;

  beforeEach(() => {
    clock = sinon.useFakeTimers();
    state = new SimpleState();
    element = 
      <Provider theState={state}>
        <Wrapper />
      </Provider>;
  });

  afterEach(() => {
    clock.restore();
  });

  it('renders as expected', () => {
    const component = mount(element);
    expect(component).toMatchSnapshot();
  });

  it('should render null until the promise resolves for load times < 500ms', () => {
    state.delay = 200;
    const component = mount(element);
    return new Promise((resolve) => {
      expect(component.html()).toMatchSnapshot();
      clock.tick(state.delay);
      resolve();
    }).then(() => {
      expect(component.html()).toMatchSnapshot();
      clock.tick(5000);
    }).then(() => {
      expect(component.html()).toMatchSnapshot();
    });
  });

  it('should render null for 500ms, then loading until the promise resolves for load times > 500ms', () => {
    state.delay = 2000;
    const component = mount(element);
    return new Promise((resolve) => {
      expect(component.html()).toMatchSnapshot();
      clock.tick(500);
      resolve();
    }).then(() => {
      expect(component.html()).toMatchSnapshot();
      clock.tick(state.delay - 500);
    }).then(() => {
      expect(component.html()).toMatchSnapshot();
    });
  });

  it('should render the error message if the promise is rejected', () => {
    state.delay = 1000;
    state.shouldReject = true;
    const component = mount(element);
    // use promises to allow callbacks in loading to run between ticks
    return new Promise((resolve) => {
      expect(component.html()).toMatchSnapshot();
      clock.tick(500);
      resolve();
    }).then(() => {
      expect(component.html()).toMatchSnapshot();
      clock.tick(state.delay - 500);
    }).then(() => {
      // we need an extra then for the reject to be caught
    })
    .then(() => {
      expect(component.html()).toMatchSnapshot();
    });
  });

  it('should inject the specified store', () => {
    const component = mount(element);
    return new Promise((resolve) => {
      clock.tick(0);
      resolve();
    }).then(() => {
      component.update();
      expect(component.find('SimpleComponent').first().prop('theState')).toBe(state);
    });
  });

  it('should make the wrapped component an observer', () => {
    const component = mount(element);
    return new Promise((resolve) => {
      clock.tick(0);
      resolve();
    }).then(() => {
      component.update();
      state.message = "updated";
      expect(component.html()).toMatchSnapshot();
    });
  });

});
