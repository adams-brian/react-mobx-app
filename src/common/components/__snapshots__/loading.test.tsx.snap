// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading renders as expected 1`] = `
<Provider
  theState={
    SimpleState {
      "delay": 0,
      "message": "",
      "shouldReject": false,
    }
  }
>
  <inject-loading(SimpleComponent)-with-theState>
    <loading(SimpleComponent)
      theState={
        SimpleState {
          "delay": 0,
          "message": "",
          "shouldReject": false,
        }
      }
    />
  </inject-loading(SimpleComponent)-with-theState>
</Provider>
`;

exports[`loading should make the wrapped component an observer 1`] = `"<div>message: updated</div>"`;

exports[`loading should render null for 500ms, then loading until the promise resolves for load times > 500ms 1`] = `null`;

exports[`loading should render null for 500ms, then loading until the promise resolves for load times > 500ms 2`] = `"<div>Loading...</div>"`;

exports[`loading should render null for 500ms, then loading until the promise resolves for load times > 500ms 3`] = `"<div>message: loaded</div>"`;

exports[`loading should render null until the promise resolves for load times < 500ms 1`] = `null`;

exports[`loading should render null until the promise resolves for load times < 500ms 2`] = `"<div>message: loaded</div>"`;

exports[`loading should render null until the promise resolves for load times < 500ms 3`] = `"<div>message: loaded</div>"`;

exports[`loading should render the error message if the promise is rejected 1`] = `null`;

exports[`loading should render the error message if the promise is rejected 2`] = `"<div>Loading...</div>"`;

exports[`loading should render the error message if the promise is rejected 3`] = `"<div>Error: error message</div>"`;
